src/lm/language_model.cpp

```diff
@@ -7,213 +7,225 @@
         }
-    }
 
-    // turn counts into probabilities
-    for (auto& map : dist_)
-    {
-        double sum = 0.0;
-        for (auto& end : map.second)
-            sum += end.second;
-        for (auto& end : map.second)
-            end.second /= sum;
+        auto first_tab = buffer.find_first_of('\t');
+        float prob = std::stof(buffer.substr(0, first_tab));
+        auto second_tab = buffer.find_first_of('\t', first_tab + 1);
+        auto ngram = buffer.substr(first_tab + 1, second_tab - first_tab - 1);
+        float backoff = 0.0;
+        if (second_tab != std::string::npos)
+            backoff = std::stof(buffer.substr(second_tab + 1));
+        lm_[N_].insert(ngram, prob, backoff);
+
+        if (N_ == 0)
+            unigrams << ngram << std::endl;
     }
+
+    ++N_;
 }
 
-std::string language_model::next_token(const std::deque<std::string>& tokens,
-                                       double random) const
+std::vector<std::pair<std::string, float>>
+    language_model::top_k(const sentence& prev, size_t k) const
 {
-    auto str = make_string(tokens);
-    auto it = dist_.find(str);
-    if (it == dist_.end())
-        throw std::runtime_error{"couldn't find previous n - 1 tokens: " + str};
+    // this is horribly inefficient due to this LM's structure
+    using pair_t = std::pair<std::string, float>;
+    auto comp = [](const pair_t& a, const pair_t& b)
+    {
+        return a.second > b.second;
+    };
+    std::vector<pair_t> candidates;
 
-    double cur = 0.0;
-    for (auto& end : it->second)
+    sentence candidate = prev;
+    candidate.push_back("word"); // the last item is replaced each iteration
+    for (auto& word : vocabulary_)
     {
-        cur += end.second;
-        if (cur > random)
-            return end.first;
+        auto candidate = sentence{prev.to_string() + " " + word};
+        candidates.emplace_back(word, log_prob(candidate));
+        std::push_heap(candidates.begin(), candidates.end(), comp);
+        if (candidates.size() > k)
+        {
+            std::pop_heap(candidates.begin(), candidates.end(), comp);
+            candidates.pop_back();
+        }
     }
 
-    throw std::runtime_error{"could not generate next token: " + str};
+    for (auto end = candidates.end(); end != candidates.begin(); --end)
+        std::pop_heap(candidates.begin(), end, comp);
+
+    return candidates;
 }
 
-std::string language_model::generate(unsigned int seed) const
+void language_model::load_vocab()
 {
-    std::default_random_engine gen(seed);
-    std::uniform_real_distribution<double> rdist(0.0, 1.0);
-
-    // start generating at the beginning of a sequence
-    std::deque<std::string> ngram;
-    for (size_t n = 1; n < N_; ++n)
-        ngram.push_back("<s>");
-
-    // keep generating until we see </s>
-    std::string output;
-    std::string next = next_token(ngram, rdist(gen));
-    while (next != "</s>")
+    std::string word;
+    std::ifstream unigrams{prefix_ + "0.strings"};
```

Comment:
It's relatively small, and for debugging when I was writing it.

But.. it also allows a nice hack for `diff` if I set the function words to point to this file. Then I can insert all words in the vocab easily. (Since it's a user-specified file, it should be human-readable...)